### ISSUE TITLE
ILLink is removing branch in System.SR.GetResourceString

### ISSUE THREAD

**Issue by eerhardt** [mono/linker#1112]
In the `dotnet/runtime` libraries, we have the following code in [System.SR](https://github.com/dotnet/runtime/blob/4b85b6d5fd1ce169183440d5e9d3a1d48594ad7e/src/libraries/Common/src/System/SR.cs#L13-L27):

```C#
        // This method is used to decide if we need to append the exception message parameters to the message when calling SR.Format.
        // by default it returns false.
        // Native code generators can replace the value this returns based on user input at the time of native code generation.
        // Marked as NoInlining because if this is used in an AoT compiled app that is not compiled into a single file, the user
        // could compile each module with a different setting for this. We want to make sure there's a consistent behavior
        // that doesn't depend on which native module this method got inlined into.
        [MethodImpl(MethodImplOptions.NoInlining)]
        private static bool UsingResourceKeys() => false;

        internal static string GetResourceString(string resourceKey, string? defaultString = null)
        {
            if (UsingResourceKeys())
            {
                return defaultString ?? resourceKey;
            }

            // do the resource look-up
```

In the `3.1` libraries, the above code would produce the following IL:

```
.method assembly hidebysig static string 
        GetResourceString(string resourceKey,
                          [opt] string defaultString) cil managed
{
  .param [2] = nullref
  // Code size       49 (0x31)
  .maxstack  2
  .locals (string V_0)
  IL_0000:  call       bool System.SR::UsingResourceKeys()
  IL_0005:  brfalse.s  IL_000e
  IL_0007:  ldarg.1
  IL_0008:  dup
  IL_0009:  brtrue.s   IL_000d
  IL_000b:  pop
  IL_000c:  ldarg.0
  IL_000d:  ret
  IL_000e:  ldnull
  IL_000f:  stloc.0
  .try
  {
    IL_0010:  call       class [System.Resources.ResourceManager]System.Resources.ResourceManager System.SR::get_ResourceManager()
```

However, in the latest `5.0` builds, this is the IL:

```
.method assembly hidebysig static string 
        GetResourceString(string resourceKey,
                          [opt] string defaultString) cil managed
{
  .param [2] = nullref
  // Code size       42 (0x2a)
  .maxstack  2
  .locals (string V_0)
  IL_0000:  call       bool System.SR::UsingResourceKeys()
  IL_0005:  brfalse.s  IL_0007
  IL_0007:  ldnull
  IL_0008:  stloc.0
  .try
  {
    IL_0009:  call       class [System.Runtime]System.Resources.ResourceManager System.SR::get_ResourceManager()
```

Note that we are still issuing the call to `System.SR::UsingResourceKeys()`, but we are stripping out the branch code that will `return defaultString ?? resourceKey;`.

The above SR code is what we have in place for being able to remove resource strings on UWP. I was attempting to flip the hard-coded bool to `true` to test out how much size we could save if we used the same approach. However, since ILLink is running during our library build (before we try trimming for size in the context of an app), it is causing the pattern to no longer work.

Maybe ILLink should respect the `[MethodImpl(MethodImplOptions.NoInlining)]` attribute when it looks for constant method bodies, and not try to fold them?

cc @marek-safar @vitek-karas @MichalStrehovsky @sbomer

**Comment by marek-safar** [comment]
That's by design where the default constant propagation kicks-in. You can use the following to disable
- disable the optimization for this assembly
- add `IntrinsicAttribute` to the method for the linker to not analyze the body
- change the body to return static field instead of constant

**Comment by eerhardt** [comment]
> add IntrinsicAttribute to the method for the linker to not analyze the body

> We thought about it but decided not to as the [MethodImpl(MethodImplOptions.NoInlining)] should be there for a bit different purposes.

I don't know if the IntrinsicAttribute is the correct one to mark this as "don't inspect my body". That one is a totally different purpose - it means the runtime is going to replace the body of the method.

https://github.com/dotnet/runtime/blob/2c87db6b2ec2d4a7f0c20b33378b0f7aaeccea26/src/libraries/System.Private.CoreLib/src/System/Runtime/CompilerServices/IntrinsicAttribute.cs#L7-L9

```
    // Calls to methods or references to fields marked with this attribute may be replaced at
    // some call sites with jit intrinsic expansions.
    // Types marked with this attribute may be specially treated by the runtime/compiler.
```

To me, marking `[MethodImpl(MethodImplOptions.NoInlining)]` makes more sense because:

1. The [description ](https://docs.microsoft.com/en-us/dotnet/api/system.runtime.compilerservices.methodimploptions?view=netframework-4.8#fields) of this attribute fits with this scenario: `The method cannot be inlined. Inlining is an optimization by which a method call is replaced with the method body.` That's exactly what the linker is trying to do here - inline the constant result.
2. `MethodImpl` is public, so people outside of `CoreLib` can use it. `System.Console` can't use the IntrinsicAttribute as it exists today, and neither can 3rd parties.

**Comment by eerhardt** [comment]
A separate issue is that we should be removing the call to the method `UsingResourceKeys`, but we aren't. I've logged #1113 specifically for that.

**Comment by marek-safar** [comment]
> I was attempting to flip the hard-coded bool to true to test out how much size we could save if we used the same approach

My recommendation was for this your request. If you ask for a permanent solution then IntrinsicAttribute is not the way forward.

> That's exactly what the linker is trying to do here - inline the constant result.

That's exactly what linker is *NOT* doing. The linker does not inline the call as you can see in the output IL.

**Comment by vitek-karas** [comment]
Per offline discussion - some more reasons as to why is this happening: https://github.com/mono/linker/issues/1106#issuecomment-614910683

**Comment by MichalStrehovsky** [comment]
In what scenarios is inlining the return value of a NoInlining method beneficial? Do we have a use case where this makes a meaningful difference (or any difference at all)?

**Comment by marek-safar** [comment]
We could fix linker to ignore NoInlining as well but no inlining should not alter program behaviour/result which is what is happening in this case. I still think the framework code is abusing NoInlinig in this case for something that `IntrinsicAttribute` is used elsewhere in the libraries for.

**Comment by MichalStrehovsky** [comment]
We could mark it `Intrinsic`, but `Intrinsic` is typically only used in CoreLib. This affects all SR classes within framework (so e.g. System.Linq would need to carry the definition of the IntrinsicsAttribute so that it can put it on the method - because the attribute is not public). I'm not opposed to doing that, but I expect some pushback from the libraries team if we were to do that.

**Comment by vargaz** [comment]
This really constraints the ability of products to use substitutions since by the time they receive the IL, it's already substituted.
Also, the linker does inline this case, since it uses the result value of the method to rewrite the caller. 'do not inline' usually includes 'do not do cross function optimization' in most languages, like in C/C++. So I'd suggest the linker should do the same.

**Comment by vargaz** [comment]
It would also be nice to be able to turn off this behavior by an option and just do the substitutions in the xml file.

**Comment by marek-safar** [comment]
> It would also be nice to be able to turn off this behavior by an option and just do the substitutions in the xml file.

That's not necessary. We could simply not to apply the attribute for targets/RIDs where we don't want the "dynamic" behaviour so the substitutions would still kick-in.

**Comment by vargaz** [comment]
Having the ability to turn off optimizations is a very useful feature, it can give users a workaround if they run into a bug with the optimization.

**Comment by marek-safar** [comment]
There is the switch `--disable-opt NAME [ASM]` to turn it off globally or on assembly-level just don't think it's useful for this scenario

**Comment by eerhardt** [comment]
> I still think the framework code is abusing NoInlinig in this case for something that IntrinsicAttribute is used elsewhere in the libraries for.

I disagree with this. See my comment: https://github.com/mono/linker/issues/1112#issuecomment-614918277. As @MichalStrehovsky points out above, the `Intrinsic` attribute is only used in CoreLib, since it is tied to the runtime. It is not used in any of the other libraries. This attribute is a way for the JIT to recognize certain methods that can be replaced. See the discussion when it was introduced: https://github.com/dotnet/coreclr/pull/13815.

>> That's exactly what the linker is trying to do here - inline the constant result.

> That's exactly what linker is NOT doing. The linker does not inline the call as you can see in the output IL.

It might not be inlining the whole method, but it _is_ taking the constant result of the method and inlining that result to remove branches in the callsite. In my mind, that is still a form of "inlining".

**Comment by marek-safar** [comment]
@eerhardt I had another look at this and I'm still not sure we want NoInlining on this method as we ideally want to inline this method on some workloads. So if we go with NoInling we'll have to do additional work to remove NoInlining after linker is done.

What about fixing this on libraries side to work with all runtimes and linker configs as following (no NoInlining involved).

```c#
        private static bool UsingResourceKeys() => HasResourceStrings();

        [MethodImpl(MethodImplOptions.InternalCall)]
        private static extern bool HasResourceStrings();
```

This will also work perfectly fine with substitutions when needed.

**Comment by vitek-karas** [comment]
Just adding my two cents:
I understand that in some scenarios (desktop apps for example) we might be able to inline the call, I don't see that as very important though. This code is not called from very performance sensitive code paths (definitely not in a loop or similar), so inlining the calls is likely not important for perf.

In my mind marking the method as non-inlinable makes perfect sense as we indicate that additional tools may decide to overwrite the method's body (which is exactly what we do with feature switches/substitutions).

As for the proposed `InternalCall` solution - I don't see how that would work. If we don't run linker on the code at all, that internal call would remain there. So runtime would go looking for an intrinsic or something similar which we would have to provide somehow. I would very much prefer if we had a solution which is managed code only.

**Comment by eerhardt** [comment]
Could we use the feature switch design? https://github.com/dotnet/designs/blob/master/accepted/2020/feature-switch.md

Basically change:

```C#
[MethodImpl(MethodImplOptions.NoInlining)]
private static bool UsingResourceKeys() => false;
```

to 

```C#
private static bool UsingResourceKeys { get; } => InitializeUsingResourceKey();

private static bool InitializeUsingResourceKey()
{
    if (!AppContext.TryGetSwitch("System.SR.UsingResourceKeys", out bool ret))
    {
        ret = false;
    }
    return ret;
}
```

**Comment by marek-safar** [comment]
> Could we use the feature switch design

No problem with that and it might actually be useful when debugging a configuration which has broken core functionality.

**Comment by vitek-karas** [comment]
That's a great idea @eerhardt. 

Just observation: should we actually mark the `InitializeUsingResourceKey` as `NoInlining`? Technically the linker won't be able to correctly apply the feature switch if the method is for some reason inlined before it gets to the linker.

**Comment by eerhardt** [comment]
I don't think `InitializeUsingResourceKey()` needs to be marked as `NoInlining`. My understanding of the feature switch design is that the linker is going to be replacing the value of the `private static bool UsingResourceKeys` property. So we should be able to ignore `InitializeUsingResourceKey()` altogether.

**Comment by vitek-karas** [comment]
You're right, of course - not sure what I was using, but definitely not my brain....

**Comment by eerhardt** [comment]
> but definitely not my brain....

LOL. I'm not sure I'd go that far.... 😉

**Comment by MichalStrehovsky** [comment]
> So we should be able to ignore InitializeUsingResourceKey() altogether.

Thinking about this (maybe slightly unrelated) -  would the feature switch leave behind the `InitializeUsingResourceKey` method, along with a call to it from the class constructor? Not that it matters much in this case, just thinking whether that's preventable. Leaving dead code and a dead field around might result in an odd debugging experience.

**Comment by eerhardt** [comment]
My thinking was that no one would be reading that backing field. It will only be set in the static constructor. So my expectation is that the linker should be able to remove the field safely.

Once the field is removed, then there is no reason to call the method from the static constructor in this case. But that might not always be true for all "Initialize" methods, say if they mutate some other state. We had the `[Pure]` attribute for things like this in the past, but [we've been removing them](https://github.com/dotnet/runtime/pull/35118#discussion_r410679735).

**Comment by MichalStrehovsky** [comment]
> Once the field is removed, then there is no reason to call the method from the static constructor in this case. But that might not always be true for all "Initialize" methods, say if they mutate some other state. We had the [Pure] attribute for things like this in the past, but we've been removing them.

Yeah, `[Pure]` doesn't have the sort of guarantees around side effects that linker could rely on - nobody checks there are indeed no side effects.

The leftover code is unfortunate because besides being observable in debugger/reflection it prevents a bunch of linking from happening (e.g. if this is the only thing in the cctor, we could have removed the cctor; same for `AppContext` and implementation details it calls into). In this sense the [previous design](https://github.com/dotnet/designs/pull/42) allowed a cleaner excision of the feature.

**Comment by marek-safar** [comment]
> It will only be set in the static constructor. So my expectation is that the linker should be able to remove the field safely.

This is quite tricky and the linker is not able to do this yet. If we are concern about the extra field and static ctor in this case we need to move the field caching to `InitializeUsingResourceKey`.

**Comment by vitek-karas** [comment]
We can get half-way there with substitution XML though. Just like the feature switch will set the constant value for the property getter, it could also remove the body of the `InitializeUsingResourceKey`:
```XML
      <method signature="InitializeUsingResourceKey()" body="remove" />
```

This will remove the body of the method and replace it with a simple throw. So we're still left with the method and so on, but its body is gone and all it depends on potentially as well.
We would use the same feature switch just like for the constant return value.

Going forward I think we could also introduce `callsite="remove"` (or something like that) which would also potentially try to erase all call sites.

**Comment by MichalStrehovsky** [comment]
> it could also remove the body of the `InitializeUsingResourceKey`

If we still call that method from the class constructor, this would cause a `TypeInitializationException` to be thrown, right?

> Going forward I think we could also introduce callsite="remove" (or something like that) which would also potentially try to erase all call sites.

If I have `DoSomething(SomethingElse())` with `callsite="remove"` for `DoSomething`, we're still going to leave the call in for `SomethingElse()`, right?

**Comment by vitek-karas** [comment]
I don't know what it should do in all cases, I guess what I'm saying is that we can basically introduce what you had in your original proposal through substitution.xml as well... or something similar which would work for us.

**Comment by eerhardt** [comment]
If we changed the "feature switch" design from replacing the `private static bool` property to instead replacing the body of `InitializeUsingResourceKey()` with `return true/false;`, that would address the concern, right?

That could potentially solve the "how do we map the AppContext switch name (i.e. "System.SR.UsingResourceKeys") with the property name?" issue as well. If the linker recognized the pattern: `AppContext.TryGetSwitch("System.SR.UsingResourceKeys", out bool ret)` when it saw the `System.SR.UsingResourceKeys` feature switch being set, it could just rewrite `AppContext.TryGetSwitch("System.SR.UsingResourceKeys", out bool ret)` to `bool ret = true/false;` and rewrite the return of `TryGetSwtich` to be `true`. Then we don't need a mapping, and everything else should fall away as appropriate.

**Comment by MichalStrehovsky** [comment]
Yep, intrinsically recognizing `AppContext.TryGetSwitch` would work too, and it would be the ideal user experience. But it means a bit more linker work because the pattern translates to this behind the scenes:

```
class SR
{
    static bool InitializeUsingResourceKeys()
    {
        // ... This part is mostly easy to constant propagate
    }

    // This is readonly so knowing the value after cctor runs is enough
    static readonly bool backingField_UsingResourceKeys;

    static SR()
    {
        // Linker needs to find this assignment in the class constructor
        backingField_UsingResourceKeys = InitializeUsingResourceKeys();
    }

    static bool UsingResourceKeys => backingField_UsingResourceKeys;
}
```

Basically, we have many steps until the constant reaches the property getter.

**Comment by eerhardt** [comment]
> Basically, we have many steps until the constant reaches the property getter.

But the linker shouldn't have to worry about any of those steps - the normal execution of the code will cause the backing field/property to be set correctly.

So the code gets transformed from:

## Pre-Link
```C#
class SR
{
    static bool InitializeUsingResourceKeys()
    {
        if (!AppContext.TryGetSwitch("System.SR.UsingResourceKeys", out bool ret))
        {
            ret = false;
        }
        return ret;
    }

    static readonly bool backingField_UsingResourceKeys;

    static SR()
    {
        backingField_UsingResourceKeys = InitializeUsingResourceKeys();
    }

    static bool UsingResourceKeys => backingField_UsingResourceKeys;
}
```

## Post-Link
```C#
class SR
{
    static bool InitializeUsingResourceKeys()
    {
        bool ret = true/false; // depending on the feature switch setting
        if (!true)
        {    // obviously the linker will now be able to trim this branch because of the constant,
             // but leaving it here for clarity
            ret = false;
        }
        return ret;
    }

    static readonly bool backingField_UsingResourceKeys;

    static SR()
    {
        backingField_UsingResourceKeys = InitializeUsingResourceKeys();
    }

    static bool UsingResourceKeys => backingField_UsingResourceKeys;
}
```

The only part that needs to be touched is the call to `AppContext.TryGetSwitch`. The rest can stay as initially written.

_____

One drawback to this approach is that the code in the libraries would have to follow this pattern exactly - specifically calling `TryGetSwitch` with an inline hard-coded string. We aren't following that pattern exactly today with [`GlobalizationMode.Invariant`](https://github.com/dotnet/runtime/blob/a9626b9d3e1f52234c6864cc5d9f11252fa5d3e4/src/coreclr/src/System.Private.CoreLib/src/System/Globalization/GlobalizationMode.cs#L13-L34), but we can update it to follow this pattern, if necessary. It would be a trade-off between maintaining the mapping from AppContext switch name to the IL property vs. coding following a convention.

**Comment by MichalStrehovsky** [comment]
If linker doesn't propagate the values, it wouldn't be able to eliminate the dead branches that result from switching on the bool:

```csharp
if (UsingResourceKeys)
{
    DoSomething();
}
else
{
    DoSomethingElse();
}
```

**Comment by marek-safar** [comment]
I have a feeling that we are overthinking this a bit. The additional single bool field is really not that important and if necessary there are ways how to remove it. I'd not go the route of recognizing `AppContext.TryGetSwitch` in linker though.

**Comment by vargaz** [comment]
https://github.com/mono/linker/pull/1230

**Comment by vargaz** [comment]
The inlining problem should be fixed now, now the linker will avoid inlining the UsingResourceKeys () method by default, but will still do it if it has a substitution.

**Comment by eerhardt** [comment]
Closing this issue as the original concern has been addressed with #1230.
